syncer/dispatcher_test.go

```diff
@@ -1,35 +1,82 @@
 package syncer_test
 
 import (
+	"context"
+	"strconv"
 	"testing"
 
 	"github.com/filecoin-project/go-filecoin/block"
+	"github.com/filecoin-project/go-filecoin/types"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/filecoin-project/go-filecoin/syncer"
 	tf "github.com/filecoin-project/go-filecoin/testhelpers/testflags"
 )
 
-func TestNewDispatcher(t *testing.T) {
+type fakeSyncer struct {
+	headsCalled []block.TipSetKey
+}
+
+func (fs *fakeSyncer) HandleNewTipSet(ctx context.Context, ci *block.ChainInfo, t bool) error {
+	fs.headsCalled = append(fs.headsCalled, ci.Head)
+	return nil
+}
+
+func TestDispatchStartHappy(t *testing.T) {
 	tf.UnitTest(t)
-	syncer.NewDispatcher()
+	s := &fakeSyncer{
+		headsCalled: make([]block.TipSetKey, 0),
+	}
+	testDispatch := syncer.NewDispatcher(s)
+
+	cis := []*block.ChainInfo{
+		chainInfoFromHeight(t, 0),
+		chainInfoFromHeight(t, 42),
+		chainInfoFromHeight(t, 3),
+		chainInfoFromHeight(t, 16),
+		chainInfoFromHeight(t, 2),
+	}
+
+	testDispatch.Start(context.Background())
+
+	// set up a blocking channel and register to unblock after 5 synced
+	wait := make(chan struct{})
+	done := func() {
+		wait <- struct{}{}
+	}
```

Comment:
Strongly recommend a `moresync.Latch` for this kind of thing. It can even do the counting for you.

```
allDone := moresync.NewLatch(5)
testDispatch.SetOnProcessedHandler(allDone.Done) // called at the bottom of every dispatch loop iteration
go func() { ... }
allDone.Wait()
```

You could also roll this into the fakeSyncer (which is really a mock).